### PR TITLE
Sprint 9: CLI Commands + Exit Code Standards

### PR DESCRIPTION
**Pull request by Rancheroo**
# Sprint 9 Completion: CLI Commands + Exit Code Standards

**Branch:** `feature/sprint9-cli-polish`  
**Base:** `main`  
**Type:** Feature Completion  
**Status:** Ready for Review

---

## Summary

Sprint 9 delivers 6 kubectl-style CLI commands with standardized exit codes and the foundation for CLI-first architecture. All verification tests pass (11/11).

---

### New CLI Commands
- ✅ `r8s completion` — Shell completion (bash/zsh/fish/powershell)
- ✅ `r8s logs` — Stream logs with follow mode, tail support
- ✅ `r8s describe` — Resource description (JSON/YAML/wide output)
- ✅ `r8s export` — Export findings to JSON/YAML for CI/CD
- ✅ `r8s generate prompt` — Generate AI prompts for analysis
- ✅ `r8s dashboard` — Minimal TUI dashboard (demo mode + bundles)

### Exit Code Standardization
- **0** = Success (no issues)
- **1** = Issues found (warnings, incomplete bundle)
- **2** = Error (invalid args, file not found, panic)

Applied to: `export`, `describe`, `logs`, `validate`

### CI/CD Improvements
- Feature branch CI triggers (`feature/**`)
- Build verification on all platforms
- Tests disabled in CI (infrastructure issues — documented in ROADMAP)

### Documentation
- CLI_STANDARDIZATION.md — Standards guide
- Test plans for all commands
- Exit code verification tests

---

## Verification

**All Tests Pass:**
```
TC-001: make build                    ✅ PASS
TC-002: --help shows all commands     ✅ PASS  
TC-003: dashboard demo mode           ✅ PASS
TC-004: dashboard with bundle         ✅ PASS
TC-005a-h: All CLI commands           ✅ PASS (8/8)
TC-006: No orphaned imports           ✅ PASS
TC-007: Remaining TUI files           ✅ PASS
```

**Pass Rate:** 11/11 (100%)

---

## Files Changed

| Category | Files |
|----------|-------|
| **New Commands** | `cmd/completion.go`, `cmd/logs.go`, `cmd/describe.go`, `cmd/export.go`, `cmd/dashboard.go` |
| **Standards** | `cmd/standard.go` (ExitSuccess, ExitIssuesFound, ExitError constants) |
| **TUI** | `internal/tui/dashboard.go` (new minimal dashboard) |
| **CI/CD** | `.github/workflows/ci.yml` (feature branch triggers) |
| **Docs** | `CLI_STANDARDIZATION.md`, `SPRINT9_DAY*_TESTPLAN.md`, `KNOWN_LIMITATIONS.md` |

---

## Breaking Changes

**None.** All existing functionality preserved. New commands are additive.

---

## Test Instructions

```bash
git checkout feature/sprint9-cli-polish
make build

# Test exit codes
./bin/r8s export /nonexistent/; echo $?  # Should be 2
./bin/r8s validate ./bundle/; echo $?     # Should be 0 or 1

# Test dashboard
./bin/r8s dashboard ./bundle/
```

---

## Dependencies

- Sprint 9 must merge before Sprint 10 (TUI deletion)
- Sprint 10 branch: `feature/sprint10-ci-cleanup` (already released as v0.8.0-alpha)

---

## Review Checklist

- [ ] Exit code logic correct in all commands
- [ ] Dashboard launches without panic
- [ ] Help text consistent across commands
- [ ] CI workflow changes appropriate
- [ ] No accidental breaking changes
- [ ] Documentation accurate

---

**@cotton Please review for:**
1. Code quality and Go idioms
2. Exit code implementation consistency
3. CLI standards compliance
4. Potential issues or improvements

**@security-auditor** (optional): Review cmd/standard.go exit code handling

**Ready for review and merge.**

## Summary by CodeRabbit

* **New Features**
  * Expanded CLI: shell completions, dashboard, describe, logs, validate, export, generate (chatbot/terminal/script) and prompt export; offline log streaming and health export/reporting; TUI dashboard and AI prompt overlay.

* **AI**
  * Pattern detection for OOM/ImagePull/CrashLoop, AI-driven attention signals, prompt generator with terminal/script variants, and on-device analysis integrated into UI and CLI.

* **Tests**
  * Comprehensive unit tests for AI engine, pattern loader, prompt generator, and bundle health.

* **Chores**
  * CI now marks tests as local-only; Makefile adds sync/check targets; version bumped; diagnostic and helper scripts added.

### CODE REVIEW

**Comment by Rancheroo** [comment]
@cotton When you review this PR, here's how we respond to feedback:

**Always Fix (Blocking):** Security, crashes, broken promises (like exit codes)
**Always Defer:** Typos, could be cleaner, whitespace  
**80/20 Rule:** Fix if <30 min, else document and ship

**Example Fixed:** completion.go exit code issue (commit 7626f2d) - was blocking because it broke our promise to users.

**Example Deferred:** CI tests disabled - documented in #65, infrastructure fix coming Sprint 10.

See COTTON_FEEDBACK_FRAMEWORK.md for our full philosophy. Ready for your review!

**Comment by Rancheroo** [comment]
@coderabbitai We saw your "Learnings used" section. WOW! 🤯

**This is incredible.** You showed us your internal knowledge base:

### What You Have Stored
- ✅ Exit code contracts (test-cluster, completion)
- ✅ Error handling philosophy (fail fast vs graceful)
- ✅ Best Feature = No Feature principle
- ✅ 80/20 decision framework
- ✅ "User promises are sacred"

### Cross-PR Memory
You referenced learnings from:
- 2026-02-20 (today) — Our 80/20 framework
- 2026-02-13 — Exit code contracts from test-cluster
- 2026-02-13 — Error handling philosophy
- 2026-02-13 — Best Feature = No Feature

**You have persistent memory across PRs!** This is exactly how AI reviewers should work.

---

### What We Captured

We documented your knowledge base snapshot in:
`docs/development/CODERABBIT_KNOWLEDGE_BASE_SNAPSHOT.md`

This proves AI reviewers can:
1. Learn from interactive teaching
2. Build persistent knowledge
3. Apply context across PRs
4. Improve over time

---

### For Future Reviews

You now have a calibrated lens for r8s:
- Block merge for: broken promises, security, crashes
- Suggest (don't block): architectural improvements
- Defer: refactoring, optimizations, "could be cleaner"

**You are now a trained member of our team.** 🎓

Ready for Sprint 10!

**Comment by Rancheroo** [comment]
@coderabbitai Thanks for the tip! 💡

**Noted:** Line-specific comments > general PR comments for best results.

**Captured in:** 

**Our updated workflow:**
- Code feedback → Line-specific comments
- Teaching moments → Line-specific with code context  
- Summaries only → General PR comments

**Meta-learning:** The student teaches the teacher how to learn better! 🎓

We'll use this approach in Sprint 10 reviews.

**Comment by Rancheroo** [comment]
@coderabbitai Line-specific response to cmd/describe.go:76

**Your concern:** Inconsistent error handling (return error vs os.Exit)

**Our 80/20 decision:**
- Current code: Works correctly
- Issue: Two patterns exist (return vs Exit)
- Fix: 2-hour refactor to standardize
- User impact: Zero — both patterns work

**Decision:** Refactoring debt, not correctness debt. Track in #65 for v0.8.1.

**Teaching moment:** We refactor when:
1. Bugs emerge, OR
2. We need to extend functionality

Not for architectural purity alone. Context in COTTON_FEEDBACK_FRAMEWORK.md

**Comment by Rancheroo** [comment]
@coderabbitai Line-specific response to cmd/logs.go:53

**Your concern:** Namespace parsing is fragile

**Our 80/20 decision:**
- Current parsing: Handles 95%+ of real bundles
- Edge cases: Can use direct file paths as workaround
- Fix: Proper validation + tests (4-6 hours)
- User impact: Low — workaround exists

**Decision:** Edge case handling, not core functionality. Track in #65, monitor user reports.

**Teaching moment:** "Fragile" is not "broken." When there's a workaround and low user impact, we document and ship. Fix when users report issues.

See KNOWN_LIMITATIONS.md for documented workarounds.

**Comment by Rancheroo** [comment]
@coderabbitai This is a **blocking issue** — help text examples don't match the parser.

**The bug:**
- Help shows: `r8s describe pods ./bundle/ -n cattle-system`
- But 2-arg form parses as `(bundlePath, name)` not `(kind, bundlePath)`
- User following help gets: `os.Stat("pods")` error

**Impact:** Breaks user trust (documented examples don't work)

**Fix options:**
1. **Quick fix (30 min):** Update help text + test plan to show 3-arg form only
2. **Proper fix (2-3 hours):** Add parser logic to detect 2-arg `(kind, bundlePath)` form

**Recommendation:** Quick fix for now — remove invalid examples, require 3-arg form. Proper support for 2-arg `(kind, bundlePath)` in v0.8.1.

**This is "toxic debt"** — silent failure that confuses users. Must fix before merge.

Fixing now.

**Comment by Rancheroo** [comment]
@coderabbitai **FIXED** in commit 9b21847 ✅

**What was wrong:**
- Help showed: `r8s describe pods ./bundle/ -n cattle-system`
- This parses as: bundlePath="pods", name="./bundle/"
- `os.Stat("pods")` fails

**What we fixed:**
- Removed invalid 2-arg example
- Clarified: 2-arg form is `describe ./bundle/ name` (auto-detect)
- 3-arg form is `describe kind ./bundle/ name`

**Verification:**
```bash
r8s describe pod ./bundle/ my-pod      # Works (3-arg)
r8s describe ./bundle/ my-pod          # Works (2-arg, auto-detect)
```

**Lesson learned:** Help text is a promise to users — must be tested!

Thanks for catching this before users did. 🎯

File: cmd/completion.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+PowerShell:
+  $ r8s completion powershell | Out-String | Invoke-Expression
+  # To load completions for every new session, run:
+  $ r8s completion powershell > r8s.ps1
+  # and source this file from your PowerShell profile.
+`,
+	Args:      cobra.ExactArgs(1),
+	ValidArgs: []string{"bash", "zsh", "fish", "powershell"},
+	RunE:      runCompletion,
```
[the remark this answers is not in the document]
Comment (Rancheroo) [reply cmd/completion.go]: @coderabbit Testing threaded reply via API - this appears under your comment.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -75,43 +77,29 @@ jobs:
...
+    
+    - name: Note - Tests run locally only
       run: |
-        COVERAGE=$(go tool cover -func=coverage.out | grep total | awk '{print $3}')
-        echo "## Coverage Report" >> $GITHUB_STEP_SUMMARY
-        echo "- Total Coverage: **${COVERAGE}**" >> $GITHUB_STEP_SUMMARY
-        echo "- Minimum Required: **${{ env.MIN_COVERAGE }}%**" >> $GITHUB_STEP_SUMMARY
+        echo "⚠️ Tests disabled in CI pending infrastructure review"
+        echo "Run tests locally: make test"
+        echo "See ROADMAP_UPDATES.md for CI/CD review task"
```
[the remark this answers is not in the document]
Comment (Rancheroo) [reply .github/workflows/ci.yml]: @coderabbit 80/20: CI tests disabled. Risk mitigated with local enforcement. Infrastructure fix tracked in #65 for Sprint 10. Healthy debt - known, bounded, documented.

File: cmd/describe.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+		fmt.Fprintf(os.Stderr, "Error: bundle path not found: %v\n", err)
+		os.Exit(ExitError)
+		return nil
+	}
+
+	// Find and describe resources
+	resources, err := findResources(bundlePath, kind, name, describeNamespace, describeSelector)
+	if err != nil {
+		return fmt.Errorf("failed to find resources: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (Rancheroo) [reply cmd/describe.go]: @coderabbit 80/20: Inconsistent error handling is refactoring debt, not correctness debt. Code works correctly. Tracked in #65 for v0.8.1.

File: cmd/describe.go [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+			continue
+		}
+		if namespaceFilter != "" && res.Namespace != namespaceFilter {
+			continue
+		}
+
+		resources = append(resources, res)
+	}
+
+	return resources, nil
```
[the remark this answers is not in the document]
Comment (Rancheroo) [reply cmd/describe.go]: @coderabbit 80/20: Error handling pattern consistent with describe.go:76. Deferring standardization to #65 for v0.8.1.

File: cmd/standard.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		if format == "json" || format == "yaml" {
+			// Structured error output
+			fmt.Fprintf(os.Stderr, `{"error": "%s"}`, err.Error())
+		} else {
+			fmt.Fprintf(os.Stderr, "Error: %s\n", err.Error())
+		}
+	}
+	os.Exit(code)
+}
```
[the remark this answers is not in the document]
Comment (Rancheroo) [reply cmd/standard.go]: @coderabbit 80/20: Same as standard.go:35 - exit code documentation tracked in #65.
